includes/class-wc-google-gtag-js.php

```diff
@@ -37,498 +36,192 @@ public static function get_instance( $options = array() ) {
+						}
+					}
+				);
 			}
 		);
-
-		return wp_json_encode( $data );
 	}
 
 	/**
-	 * Returns a list of category names the product is atttributed to
+	 * Set script data for a specific event
 	 *
-	 * @param  WC_Product $product Product to generate category line for
-	 * @return string
-	 */
-	public static function product_get_category_line( $product ) {
-		$category_names = array();
-		$categories     = get_the_terms( $product->get_id(), 'product_cat' );
-
-		$variation_data = $product->is_type( 'variation' ) ? wc_get_product_variation_attributes( $product->get_id() ) : false;
-		if ( is_array( $variation_data ) && ! empty( $variation_data ) ) {
-			$categories = get_the_terms( $product->get_parent_id(), 'product_cat' );
-		}
-
-		if ( false !== $categories && ! is_wp_error( $categories ) ) {
-			foreach ( $categories as $category ) {
-				$category_names[] = $category->name;
-			}
-		}
-
-		return join( '/', $category_names );
-	}
-
-	/**
-	 * Return list name for event
+	 * @param string       $type The type of event this data is related to.
+	 * @param string|array $data The event data to add.
 	 *
-	 * @return string
+	 * @return void
 	 */
-	public static function get_list_name(): string {
-		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
-		return isset( $_GET['s'] ) ? __( 'Search Results', 'woocommerce-google-analytics-integration' ) : __( 'Product List', 'woocommerce-google-analytics-integration' );
+	public function set_script_data( string $type, $data ): void {
+		$this->script_data[ $type ] = $data;
 	}
 
 	/**
-	 * Enqueues JavaScript to build the view_item_list event
+	 * Append data to an existing script data array
 	 *
-	 * @param WC_Product $product
-	 */
-	public static function listing_impression( $product ) {
-		$event_code = self::get_event_code(
-			'view_item_list',
-			array(
-				'items' => array(
-					array(
-						'id'       => self::get_product_identifier( $product ),
-						'name'     => $product->get_title(),
-						'category' => self::product_get_category_line( $product ),
-						'list'     => self::get_list_name(),
-					),
-				),
-			)
-		);
-
-		wc_enqueue_js( $event_code );
-	}
-
-	/**
-	 * Enqueues JavaScript for select_content and add_to_cart events for the product archive
+	 * @param string       $type The type of event this data is related to.
+	 * @param string|array $data The event data to add.
 	 *
-	 * @param WC_Product $product
+	 * @return void
 	 */
-	public static function listing_click( $product ) {
-		$item = array(
-			'id'       => self::get_product_identifier( $product ),
-			'name'     => $product->get_title(),
-			'category' => self::product_get_category_line( $product ),
-			'quantity' => 1,
-		);
-
-		$select_content_event_code = self::get_event_code(
-			'select_content',
+	public function append_script_data( string $type, $data ): void {
+		$this->set_script_data(
+			$type,
 			array(
-				'items' => array( $item ),
+				...$this->script_data[ $type ] ?? [],
```

Comment:
:nail_care: I wonder if the performance of destructing and re-constructing a new array over and over, as we push products/events in the loop matters here. Maybe we can use something even shorter

```php
	public function append_script_data( string $type, $data ): void {
		if ( ! isset( $this->script_data[ $type ] ) ) {
			$this->script_data[ $type ] = array();
		}
		$this->script_data[ $type ][] = $data;
	}
```